### ISSUE TITLE
TRÅDFRI bulbs get reported without the "bluest" white

### ISSUE THREAD

**Issue by jsve** [ebaauw/homebridge-hue#176]
I'm not sure if this is correct or not, but in the interface things get a bit confusing.

When I connect to the TRÅDFRI bulbs via the TRÅDFRI gateway (and hombridge-trådfri plugin) I get a temperature wheel like this:

![img_0953](https://user-images.githubusercontent.com/9053945/31464113-b94143cc-aed0-11e7-9d00-e4d41d33c4cb.PNG)

However, through the conbee and your plugin I see a wheel like this:
![img_0952](https://user-images.githubusercontent.com/9053945/31464146-d834bf02-aed0-11e7-956f-5ac19072fcaf.PNG)

I'm not sure, but maybe if we reported the CTmin (or max) a bit outside of what the bulb can actually show, the wheel would look a bit more like what you expect (like the first image above).

**Comment by ebaauw** [comment]
> through the conbee and your plugin I see a wheel like this:

The Tradfri white spectrum bulbs support Colour Temperature values from 250 (4000K) to 454 (~2200K).  homebridge-hue sets those values as minimum and maximum allowed values for the Colour Temperature characteristic.  It's up to Apple how they display this in the Home app.

> When I connect to the TRÅDFRI bulbs via the TRÅDFRI gateway (and hombridge-trådfri plugin) I get a temperature wheel like this

It would seem that the trandfri plugin doesn't set the min and/or max allowed values.

These are the only lights I know of that don't support the (Philips Hue) default minimum of 153 (~6500K).

> I'm not sure, but maybe if we reported the CTmin (or max) a bit outside of what the bulb can actually show, the wheel would look a bit more like what you expect (like the first image above).

Then you would be selecting values that the light cannot display.  And it would mess up the UI for other HomeKit apps.

**Comment by jsve** [comment]
Yes, you are correct in the conversions from kelvin to mired. However, this breaks the interface in most of the apps i have tried. Even the web-interface for deconz shows the entire spectra, even though the light only changes color in the red parts of the spectrum. A continuous scale is itself misguiding since the bulb is only able to show maybe 5-10 different colors.

**Comment by ebaauw** [comment]
> Yes, you are correct in the conversions from kelvin to mired.

There’s no conversion.  Both ZigBee and HomeKit use mired values.

> A continuous scale is itself misguiding since the bulb is only able to show maybe 5-10 different colors.

My E27 white spectrum light reports a continuous scale from 250 to 454.  I don’t have the equipment to measure whether it indeed displays 205 different tones, but it’s more than the three settings supported out of the box by the remote.  Having said that, HomeKit does support specifying a limited list of allowed values, so if you know which 5 to 10 values, I can support only those.  I doubt whether Apple’s Home app will display the better, though.

The standard Colour Temperature was only added to HomeKit in iOS 10.3.  Philips uses/used a custom characteristic (in mired), Aurora uses/used another custom characteristic (in Kelvin).  Eve supports these two, but not (yet?) the standard one.  Home used to fake colour temperature and translate it to hue and saturation.  It still does that for colour lights that don’t support colour temperature, as does Eve.

> I'm not sure, but maybe if we reported the CTmin (or max) a bit outside of what the bulb can actually show, the wheel would look a bit more like what you expect (like the first image above).

It will.  I suppose I could add a config.json key `ignoreCTMinMax` and use the default Colour Temperature range from 153 to 500 when that’s set.  I’m not too thrilled with yet another key, though.

**Comment by jsve** [comment]
I can see that you don't want to add more keys for this. Might be an idea to wait for a couple of weeks and see if anything gets updated from apple. On the other hand; there are a lot of flags already, so whats one more ;)

Displaying the fixed values would probably be more confusing, and I don't think that would solve anything. This is the same for all light I guess. Except the color-bulbs maybe. That they don't support a continuous colour-range I mean.

I have measured the temperature of the lights, and it does seem like the range is actually a bit more narrow than what IKEA states. Would that change anything? I measured the reds as 2500K, not 2200K like IKEA claims. Maybe this is because the signal it receives tells it to show that, or it's just a marketing claim from Ikea.

**Comment by ebaauw** [comment]
> I have measured the temperature of the lights, and it does seem like the range is actually a bit more narrow than what IKEA states. Would that change anything?

In homebridge-hue (and I suppose in any app), I can only take into account what the light (through the gateway) reports.  There might be a lookup or transformation based on the reported value, but I cannot "see" if the lights actually displays that value.  It would seem like the IKEA lights just echo the values that were last set, unlike the Philips lights, which actually report their adjusted values.  E.g. when setting the IKEA colour bulb to `{"xy": [0, 1]}`, it reports (0, 1) for (X, Y), which is outside the CIE 1931 colour space, so not even a physical colour, whereas the Philips bulb reports its green point.

> Maybe this is because the signal it receives tells it to show that

The light receives the requested colour temperature from the gateway in Mired.  The light firmware translates that value into the relative brightness of the warmer LED(s) and the cooler LED(s), mixing the right colour temperature.  Same for colour lights, but now for red, green, and blue (and sometimes even white) LEDs.  Different manufacturers (or even different models from the same manufacturer) use different (combinations) of LEDs, resulting in different ranges of supported colours.  This is why maximum lumen is only at a particular colour (temperature), when all the LEDs are at maximum brightness.
